File: troTHU/debug_capture.py
from __future__ import annotations
import json
from datetime import datetime
from pathlib import Path
from typing import Any, Dict


SENSITIVE_KEY_PARTS = (
    "authorization",
    "cookie",
    "passwd",
    "password",
    "session",
    "token",
    "secret",
    "key",
    "chat",
)


def sanitize_debug_payload(value: Any) -> Any:
    if isinstance(value, dict):
        sanitized: Dict[str, Any] = {}
        for key, item in value.items():
            key_text = str(key)
            if any(part in key_text.lower() for part in SENSITIVE_KEY_PARTS):
                sanitized[key_text] = "[redacted]"
            else:
                sanitized[key_text] = sanitize_debug_payload(item)
        return sanitized
    if isinstance(value, list):
        return [sanitize_debug_payload(item) for item in value]
    return value


def append_debug_capture(path: Path, event: str, payload: Any) -> Path:
    path.parent.mkdir(parents=True, exist_ok=True)
    record = {
        "timestamp": datetime.now().isoformat(timespec="seconds"),
        "event": event,
        "payload": sanitize_debug_payload(payload),
    }
    with path.open("a", encoding="utf-8") as file:
        file.write(json.dumps(record, ensure_ascii=False, default=str) + "\n")
    return path

File: src_legacy/security/debug-capture.js
'use strict';

const fs = require('node:fs');
const path = require('node:path');

const SENSITIVE_KEY_PARTS = [
  'authorization', 'cookie', 'passwd', 'password',
  'session', 'token', 'secret', 'key', 'chat',
];

/**
 * @param {*} value
 * @returns {*}
 */
function sanitizeDebugPayload(value) {
  if (value && typeof value === 'object' && !Array.isArray(value)) {
    const sanitized = {};
    for (const [key, item] of Object.entries(value)) {
      const keyLower = String(key).toLowerCase();
      if (SENSITIVE_KEY_PARTS.some(part => keyLower.includes(part))) {
        sanitized[key] = '[redacted]';
      } else {
        sanitized[key] = sanitizeDebugPayload(item);
      }
    }
    return sanitized;
  }
  if (Array.isArray(value)) {
    return value.map(item => sanitizeDebugPayload(item));
  }
  return value;
}

/**
 * Append a debug capture record (JSONL) to the given file path.
 * @param {string} filePath
 * @param {string} event
 * @param {*} payload
 * @returns {string} The file path written to.
 */
function appendDebugCapture(filePath, event, payload) {
  const dir = path.dirname(filePath);
  fs.mkdirSync(dir, { recursive: true });
  const record = {
    timestamp: new Date().toISOString().replace(/\.\d+Z$/, 'Z'),
    event,
    payload: sanitizeDebugPayload(payload),
  };
  fs.appendFileSync(filePath, JSON.stringify(record) + '\n', 'utf8');
  return filePath;
}

module.exports = { sanitizeDebugPayload, appendDebugCapture, SENSITIVE_KEY_PARTS };
